### ISSUE TITLE
Http/3 Example - KeyUnavailableError

### ISSUE THREAD

**Issue by maruibin** [aiortc/aioquic#182]
When I was trying to transfer files between an http/3 client and an http/3 server (using the scirpts provided as the example), I found that the KeyUnavailableError("Decryption key is not available") was raised every once in a while, and when that happened, the data transfer rate would usually drop down significantly (say, by up to about 90%). Could you please explain what this error means and what might be the potential causes? I appreciate it a lot.

**Comment by jlaine** [comment]
I'm intrigued by this, would you have any traces for this? Are you on a network with packet loss?

**Comment by maruibin** [comment]
Thanks for your reply. Yes exactly, I was testing it over an intercontinental network and there could be potential packet loss. May I know what kind of traces you were referring to? Below I just attached the verbose output from both client and server side for your reference, and please let me know if you're interested in other kind of trace.

http3 server output during a file transfer with reduced speed of unknown reason:
```
2021-03-19 22:06:11,911 DEBUG quic [34ffef4cef0dff44] Network path ('::ffff:[dest ip addr here]', 36359, 0, 0) discovered
2021-03-19 22:06:11,912 DEBUG quic [34ffef4cef0dff44] QuicConnectionState.FIRSTFLIGHT -> QuicConnectionState.CONNECTED
2021-03-19 22:06:11,918 DEBUG quic [34ffef4cef0dff44] TLS State.SERVER_EXPECT_CLIENT_HELLO -> State.SERVER_EXPECT_FINISHED
2021-03-19 22:06:12,119 DEBUG quic [34ffef4cef0dff44] Loss detection triggered
2021-03-19 22:06:12,119 DEBUG quic [34ffef4cef0dff44] Discarding epoch Epoch.INITIAL
2021-03-19 22:06:12,120 DEBUG quic [34ffef4cef0dff44] Network path ('::ffff:[dest ip addr here]', 36359, 0, 0) validated by handshake
2021-03-19 22:06:12,131 DEBUG quic [34ffef4cef0dff44] TLS State.SERVER_EXPECT_FINISHED -> State.SERVER_POST_HANDSHAKE
2021-03-19 22:06:12,131 DEBUG quic [34ffef4cef0dff44] Discarding epoch Epoch.HANDSHAKE
2021-03-19 22:06:12,131 INFO quic [34ffef4cef0dff44] ALPN negotiated protocol h3-32
2021-03-19 22:06:12,131 DEBUG quic [34ffef4cef0dff44] Stream 2 created by peer
2021-03-19 22:06:12,131 DEBUG quic [34ffef4cef0dff44] Stream 6 created by peer
2021-03-19 22:06:12,131 DEBUG quic [34ffef4cef0dff44] Stream 10 created by peer
2021-03-19 22:06:12,132 DEBUG quic [34ffef4cef0dff44] Stream 0 created by peer
2021-03-19 22:06:12,132 INFO quic [34ffef4cef0dff44] HTTP request GET /8m_file
2021-03-19 22:06:12,318 DEBUG quic [34ffef4cef0dff44] Decryption key is not available
2021-03-19 22:06:12,319 DEBUG quic [34ffef4cef0dff44] Decryption key is not available
2021-03-19 22:06:17,335 DEBUG quic [34ffef4cef0dff44] Remote max_data raised to 2097152
2021-03-19 22:06:17,335 DEBUG quic [34ffef4cef0dff44] Stream 0 remote max_stream_data raised to 2097152
2021-03-19 22:06:19,760 DEBUG quic [34ffef4cef0dff44] Remote max_data raised to 4194304
2021-03-19 22:06:19,761 DEBUG quic [34ffef4cef0dff44] Stream 0 remote max_stream_data raised to 4194304
2021-03-19 22:06:23,186 DEBUG quic [34ffef4cef0dff44] Remote max_data raised to 8388608
2021-03-19 22:06:23,187 DEBUG quic [34ffef4cef0dff44] Stream 0 remote max_stream_data raised to 8388608
2021-03-19 22:06:28,044 DEBUG quic [34ffef4cef0dff44] Remote max_data raised to 16777216
2021-03-19 22:06:28,044 DEBUG quic [34ffef4cef0dff44] Stream 0 remote max_stream_data raised to 16777216
2021-03-19 22:06:34,873 DEBUG quic [34ffef4cef0dff44] Remote max_data raised to 33554432
2021-03-19 22:06:34,874 DEBUG quic [34ffef4cef0dff44] Stream 0 remote max_stream_data raised to 33554432
2021-03-19 22:06:34,891 INFO quic [34ffef4cef0dff44] Connection close code 0x0, reason 
2021-03-19 22:06:34,891 DEBUG quic [34ffef4cef0dff44] QuicConnectionState.CONNECTED -> QuicConnectionState.DRAINING
2021-03-19 22:06:35,571 DEBUG quic [34ffef4cef0dff44] Discarding epoch Epoch.ONE_RTT
2021-03-19 22:06:35,571 DEBUG quic [34ffef4cef0dff44] QuicConnectionState.DRAINING -> QuicConnectionState.TERMINATED
```
As you may notice here, after receiving the HTTP request there is this error - "Decryption key is not available".

The corresponding http3 client output:
```
2021-03-19 22:06:11,603 DEBUG asyncio Using selector: EpollSelector
2021-03-19 22:06:11,609 DEBUG quic [34ffef4cef0dff44] TLS State.CLIENT_HANDSHAKE_START -> State.CLIENT_EXPECT_SERVER_HELLO
2021-03-19 22:06:11,810 INFO quic [34ffef4cef0dff44] Retrying with QuicProtocolVersion.DRAFT_32
2021-03-19 22:06:11,812 DEBUG quic [34ffef4cef0dff44] TLS State.CLIENT_HANDSHAKE_START -> State.CLIENT_EXPECT_SERVER_HELLO
2021-03-19 22:06:12,012 DEBUG quic [34ffef4cef0dff44] Loss detection triggered
2021-03-19 22:06:12,018 DEBUG quic [34ffef4cef0dff44] QuicConnectionState.FIRSTFLIGHT -> QuicConnectionState.CONNECTED
2021-03-19 22:06:12,019 DEBUG quic [34ffef4cef0dff44] TLS State.CLIENT_EXPECT_SERVER_HELLO -> State.CLIENT_EXPECT_ENCRYPTED_EXTENSIONS
2021-03-19 22:06:12,019 DEBUG quic [34ffef4cef0dff44] TLS State.CLIENT_EXPECT_ENCRYPTED_EXTENSIONS -> State.CLIENT_EXPECT_CERTIFICATE_REQUEST_OR_CERTIFICATE
2021-03-19 22:06:12,020 DEBUG quic [34ffef4cef0dff44] Discarding epoch Epoch.INITIAL
2021-03-19 22:06:12,020 DEBUG quic [34ffef4cef0dff44] TLS State.CLIENT_EXPECT_CERTIFICATE_REQUEST_OR_CERTIFICATE -> State.CLIENT_EXPECT_CERTIFICATE_VERIFY
2021-03-19 22:06:12,029 DEBUG quic [34ffef4cef0dff44] TLS State.CLIENT_EXPECT_CERTIFICATE_VERIFY -> State.CLIENT_EXPECT_FINISHED
2021-03-19 22:06:12,030 DEBUG quic [34ffef4cef0dff44] TLS State.CLIENT_EXPECT_FINISHED -> State.CLIENT_POST_HANDSHAKE
2021-03-19 22:06:12,030 INFO quic [34ffef4cef0dff44] ALPN negotiated protocol h3-32
2021-03-19 22:06:12,030 INFO client New session ticket received
2021-03-19 22:06:12,030 DEBUG quic [34ffef4cef0dff44] Stream 3 created by peer
2021-03-19 22:06:12,031 DEBUG quic [34ffef4cef0dff44] Stream 7 created by peer
2021-03-19 22:06:12,031 DEBUG quic [34ffef4cef0dff44] Stream 11 created by peer
2021-03-19 22:06:12,212 DEBUG quic [34ffef4cef0dff44] Decryption key is not available
2021-03-19 22:06:12,219 DEBUG quic [34ffef4cef0dff44] Decryption key is not available
2021-03-19 22:06:12,219 DEBUG quic [34ffef4cef0dff44] Sending PING (ACK-of-ACK trigger) in packet 8
2021-03-19 22:06:12,231 DEBUG quic [34ffef4cef0dff44] Discarding epoch Epoch.HANDSHAKE
2021-03-19 22:06:17,235 DEBUG quic [34ffef4cef0dff44] Local max_data raised to 2097152
2021-03-19 22:06:17,235 DEBUG quic [34ffef4cef0dff44] Stream 0 local max_stream_data raised to 2097152
2021-03-19 22:06:19,661 DEBUG quic [34ffef4cef0dff44] Local max_data raised to 4194304
2021-03-19 22:06:19,661 DEBUG quic [34ffef4cef0dff44] Stream 0 local max_stream_data raised to 4194304
2021-03-19 22:06:23,087 DEBUG quic [34ffef4cef0dff44] Local max_data raised to 8388608
2021-03-19 22:06:23,087 DEBUG quic [34ffef4cef0dff44] Stream 0 local max_stream_data raised to 8388608
2021-03-19 22:06:27,944 DEBUG quic [34ffef4cef0dff44] Local max_data raised to 16777216
2021-03-19 22:06:27,945 DEBUG quic [34ffef4cef0dff44] Stream 0 local max_stream_data raised to 16777216
2021-03-19 22:06:34,774 DEBUG quic [34ffef4cef0dff44] Local max_data raised to 33554432
2021-03-19 22:06:34,774 DEBUG quic [34ffef4cef0dff44] Stream 0 local max_stream_data raised to 33554432
2021-03-19 22:06:34,789 INFO client Response received for GET /8m_file : 8388608 bytes in 22.8 s (2.949 Mbps)
2021-03-19 22:06:34,791 DEBUG quic [34ffef4cef0dff44] QuicConnectionState.CONNECTED -> QuicConnectionState.CLOSING
2021-03-19 22:06:35,755 DEBUG quic [34ffef4cef0dff44] Discarding epoch Epoch.ONE_RTT
2021-03-19 22:06:35,756 DEBUG quic [34ffef4cef0dff44] QuicConnectionState.CLOSING -> QuicConnectionState.TERMINATED
```
As you could see from above, the speed of this transaction is 2.949 Mbps, which is significantly slower then usual, and it happens once in about every 3 tests. Otherwise the usual speed can reach about 20~30Mbps over the same network. Please see the output of a 'usual-speed' file transfer below. 

http3 server output:
```
2021-03-19 22:14:49,736 DEBUG asyncio Using selector: EpollSelector
2021-03-19 22:14:52,819 DEBUG quic [b93fa38de016aed8] Network path ('::ffff:[dest ip here]', 55609, 0, 0) discovered
2021-03-19 22:14:52,821 DEBUG quic [b93fa38de016aed8] QuicConnectionState.FIRSTFLIGHT -> QuicConnectionState.CONNECTED
2021-03-19 22:14:52,828 DEBUG quic [b93fa38de016aed8] TLS State.SERVER_EXPECT_CLIENT_HELLO -> State.SERVER_EXPECT_FINISHED
2021-03-19 22:14:53,029 DEBUG quic [b93fa38de016aed8] Discarding epoch Epoch.INITIAL
2021-03-19 22:14:53,029 DEBUG quic [b93fa38de016aed8] Network path ('::ffff:[dest ip here]', 55609, 0, 0) validated by handshake
2021-03-19 22:14:53,039 DEBUG quic [b93fa38de016aed8] TLS State.SERVER_EXPECT_FINISHED -> State.SERVER_POST_HANDSHAKE
2021-03-19 22:14:53,039 DEBUG quic [b93fa38de016aed8] Discarding epoch Epoch.HANDSHAKE
2021-03-19 22:14:53,040 INFO quic [b93fa38de016aed8] ALPN negotiated protocol h3-32
2021-03-19 22:14:53,040 DEBUG quic [b93fa38de016aed8] Stream 2 created by peer
2021-03-19 22:14:53,040 DEBUG quic [b93fa38de016aed8] Stream 6 created by peer
2021-03-19 22:14:53,040 DEBUG quic [b93fa38de016aed8] Stream 10 created by peer
2021-03-19 22:14:53,041 DEBUG quic [b93fa38de016aed8] Stream 0 created by peer
2021-03-19 22:14:53,041 INFO quic [b93fa38de016aed8] HTTP request GET /8m_file
2021-03-19 22:14:54,261 DEBUG quic [b93fa38de016aed8] Remote max_data raised to 2097152
2021-03-19 22:14:54,261 DEBUG quic [b93fa38de016aed8] Stream 0 remote max_stream_data raised to 2097152
2021-03-19 22:14:54,468 DEBUG quic [b93fa38de016aed8] Remote max_data raised to 4194304
2021-03-19 22:14:54,468 DEBUG quic [b93fa38de016aed8] Stream 0 remote max_stream_data raised to 4194304
2021-03-19 22:14:54,682 DEBUG quic [b93fa38de016aed8] Remote max_data raised to 8388608
2021-03-19 22:14:54,683 DEBUG quic [b93fa38de016aed8] Stream 0 remote max_stream_data raised to 8388608
2021-03-19 22:14:54,896 DEBUG quic [b93fa38de016aed8] Remote max_data raised to 16777216
2021-03-19 22:14:54,897 DEBUG quic [b93fa38de016aed8] Stream 0 remote max_stream_data raised to 16777216
2021-03-19 22:14:55,144 DEBUG quic [b93fa38de016aed8] Loss detection triggered
2021-03-19 22:14:55,339 DEBUG quic [b93fa38de016aed8] Remote max_data raised to 33554432
2021-03-19 22:14:55,340 DEBUG quic [b93fa38de016aed8] Stream 0 remote max_stream_data raised to 33554432
2021-03-19 22:14:55,554 INFO quic [b93fa38de016aed8] Connection close code 0x0, reason 
2021-03-19 22:14:55,555 DEBUG quic [b93fa38de016aed8] QuicConnectionState.CONNECTED -> QuicConnectionState.DRAINING
2021-03-19 22:14:56,232 DEBUG quic [b93fa38de016aed8] Discarding epoch Epoch.ONE_RTT
2021-03-19 22:14:56,232 DEBUG quic [b93fa38de016aed8] QuicConnectionState.DRAINING -> QuicConnectionState.TERMINATED
```

http3 client output:
```
2021-03-19 22:14:52,514 DEBUG asyncio Using selector: EpollSelector
2021-03-19 22:14:52,519 DEBUG quic [b93fa38de016aed8] TLS State.CLIENT_HANDSHAKE_START -> State.CLIENT_EXPECT_SERVER_HELLO
2021-03-19 22:14:52,719 INFO quic [b93fa38de016aed8] Retrying with QuicProtocolVersion.DRAFT_32
2021-03-19 22:14:52,721 DEBUG quic [b93fa38de016aed8] TLS State.CLIENT_HANDSHAKE_START -> State.CLIENT_EXPECT_SERVER_HELLO
2021-03-19 22:14:52,922 DEBUG quic [b93fa38de016aed8] Loss detection triggered
2021-03-19 22:14:52,929 DEBUG quic [b93fa38de016aed8] QuicConnectionState.FIRSTFLIGHT -> QuicConnectionState.CONNECTED
2021-03-19 22:14:52,930 DEBUG quic [b93fa38de016aed8] TLS State.CLIENT_EXPECT_SERVER_HELLO -> State.CLIENT_EXPECT_ENCRYPTED_EXTENSIONS
2021-03-19 22:14:52,930 DEBUG quic [b93fa38de016aed8] TLS State.CLIENT_EXPECT_ENCRYPTED_EXTENSIONS -> State.CLIENT_EXPECT_CERTIFICATE_REQUEST_OR_CERTIFICATE
2021-03-19 22:14:52,931 DEBUG quic [b93fa38de016aed8] Discarding epoch Epoch.INITIAL
2021-03-19 22:14:52,931 DEBUG quic [b93fa38de016aed8] TLS State.CLIENT_EXPECT_CERTIFICATE_REQUEST_OR_CERTIFICATE -> State.CLIENT_EXPECT_CERTIFICATE_VERIFY
2021-03-19 22:14:52,939 DEBUG quic [b93fa38de016aed8] TLS State.CLIENT_EXPECT_CERTIFICATE_VERIFY -> State.CLIENT_EXPECT_FINISHED
2021-03-19 22:14:52,940 DEBUG quic [b93fa38de016aed8] TLS State.CLIENT_EXPECT_FINISHED -> State.CLIENT_POST_HANDSHAKE
2021-03-19 22:14:52,940 INFO quic [b93fa38de016aed8] ALPN negotiated protocol h3-32
2021-03-19 22:14:52,940 INFO client New session ticket received
2021-03-19 22:14:52,940 DEBUG quic [b93fa38de016aed8] Stream 3 created by peer
2021-03-19 22:14:52,941 DEBUG quic [b93fa38de016aed8] Stream 7 created by peer
2021-03-19 22:14:52,941 DEBUG quic [b93fa38de016aed8] Stream 11 created by peer
2021-03-19 22:14:53,120 DEBUG quic [b93fa38de016aed8] Decryption key is not available
2021-03-19 22:14:53,141 DEBUG quic [b93fa38de016aed8] Discarding epoch Epoch.HANDSHAKE
2021-03-19 22:14:54,163 DEBUG quic [b93fa38de016aed8] Local max_data raised to 2097152
2021-03-19 22:14:54,163 DEBUG quic [b93fa38de016aed8] Stream 0 local max_stream_data raised to 2097152
2021-03-19 22:14:54,370 DEBUG quic [b93fa38de016aed8] Local max_data raised to 4194304
2021-03-19 22:14:54,370 DEBUG quic [b93fa38de016aed8] Stream 0 local max_stream_data raised to 4194304
2021-03-19 22:14:54,584 DEBUG quic [b93fa38de016aed8] Local max_data raised to 8388608
2021-03-19 22:14:54,585 DEBUG quic [b93fa38de016aed8] Stream 0 local max_stream_data raised to 8388608
2021-03-19 22:14:54,790 DEBUG quic [b93fa38de016aed8] Local max_data raised to 16777216
2021-03-19 22:14:54,790 DEBUG quic [b93fa38de016aed8] Stream 0 local max_stream_data raised to 16777216
2021-03-19 22:14:54,809 DEBUG quic [b93fa38de016aed8] Sending PING (ACK-of-ACK trigger) in packet 640
2021-03-19 22:14:54,818 DEBUG quic [b93fa38de016aed8] Sending PING (ACK-of-ACK trigger) in packet 648
2021-03-19 22:14:54,828 DEBUG quic [b93fa38de016aed8] Sending PING (ACK-of-ACK trigger) in packet 656
2021-03-19 22:14:54,837 DEBUG quic [b93fa38de016aed8] Sending PING (ACK-of-ACK trigger) in packet 664
2021-03-19 22:14:54,847 DEBUG quic [b93fa38de016aed8] Sending PING (ACK-of-ACK trigger) in packet 672
2021-03-19 22:14:54,857 DEBUG quic [b93fa38de016aed8] Sending PING (ACK-of-ACK trigger) in packet 680
2021-03-19 22:14:54,866 DEBUG quic [b93fa38de016aed8] Sending PING (ACK-of-ACK trigger) in packet 688
2021-03-19 22:14:54,876 DEBUG quic [b93fa38de016aed8] Sending PING (ACK-of-ACK trigger) in packet 696
2021-03-19 22:14:54,885 DEBUG quic [b93fa38de016aed8] Sending PING (ACK-of-ACK trigger) in packet 704
2021-03-19 22:14:54,895 DEBUG quic [b93fa38de016aed8] Sending PING (ACK-of-ACK trigger) in packet 712
2021-03-19 22:14:54,905 DEBUG quic [b93fa38de016aed8] Sending PING (ACK-of-ACK trigger) in packet 720
2021-03-19 22:14:54,914 DEBUG quic [b93fa38de016aed8] Sending PING (ACK-of-ACK trigger) in packet 728
2021-03-19 22:14:54,924 DEBUG quic [b93fa38de016aed8] Sending PING (ACK-of-ACK trigger) in packet 736
2021-03-19 22:14:54,934 DEBUG quic [b93fa38de016aed8] Sending PING (ACK-of-ACK trigger) in packet 744
2021-03-19 22:14:54,943 DEBUG quic [b93fa38de016aed8] Sending PING (ACK-of-ACK trigger) in packet 752
2021-03-19 22:14:54,954 DEBUG quic [b93fa38de016aed8] Sending PING (ACK-of-ACK trigger) in packet 760
2021-03-19 22:14:54,963 DEBUG quic [b93fa38de016aed8] Sending PING (ACK-of-ACK trigger) in packet 768
2021-03-19 22:14:54,985 DEBUG quic [b93fa38de016aed8] Sending PING (ACK-of-ACK trigger) in packet 776
2021-03-19 22:14:55,003 DEBUG quic [b93fa38de016aed8] Sending PING (ACK-of-ACK trigger) in packet 784
2021-03-19 22:14:55,013 DEBUG quic [b93fa38de016aed8] Sending PING (ACK-of-ACK trigger) in packet 792
2021-03-19 22:14:55,018 DEBUG quic [b93fa38de016aed8] Received PING (probe) response
2021-03-19 22:14:55,019 DEBUG quic [b93fa38de016aed8] Received PING (probe) response
2021-03-19 22:14:55,027 DEBUG quic [b93fa38de016aed8] Received PING (probe) response
2021-03-19 22:14:55,037 DEBUG quic [b93fa38de016aed8] Received PING (probe) response
2021-03-19 22:14:55,047 DEBUG quic [b93fa38de016aed8] Received PING (probe) response
2021-03-19 22:14:55,056 DEBUG quic [b93fa38de016aed8] Received PING (probe) response
2021-03-19 22:14:55,066 DEBUG quic [b93fa38de016aed8] Received PING (probe) response
2021-03-19 22:14:55,076 DEBUG quic [b93fa38de016aed8] Received PING (probe) response
2021-03-19 22:14:55,084 DEBUG quic [b93fa38de016aed8] Sending PING (ACK-of-ACK trigger) in packet 800
2021-03-19 22:14:55,090 DEBUG quic [b93fa38de016aed8] Received PING (probe) response
2021-03-19 22:14:55,095 DEBUG quic [b93fa38de016aed8] Sending PING (ACK-of-ACK trigger) in packet 808
2021-03-19 22:14:55,105 DEBUG quic [b93fa38de016aed8] Sending PING (ACK-of-ACK trigger) in packet 816
2021-03-19 22:14:55,109 DEBUG quic [b93fa38de016aed8] Received PING (probe) response
2021-03-19 22:14:55,109 DEBUG quic [b93fa38de016aed8] Received PING (probe) response
2021-03-19 22:14:55,115 DEBUG quic [b93fa38de016aed8] Sending PING (ACK-of-ACK trigger) in packet 824
2021-03-19 22:14:55,117 DEBUG quic [b93fa38de016aed8] Received PING (probe) response
2021-03-19 22:14:55,125 DEBUG quic [b93fa38de016aed8] Sending PING (ACK-of-ACK trigger) in packet 832
2021-03-19 22:14:55,127 DEBUG quic [b93fa38de016aed8] Received PING (probe) response
2021-03-19 22:14:55,135 DEBUG quic [b93fa38de016aed8] Sending PING (ACK-of-ACK trigger) in packet 840
2021-03-19 22:14:55,136 DEBUG quic [b93fa38de016aed8] Received PING (probe) response
2021-03-19 22:14:55,144 DEBUG quic [b93fa38de016aed8] Sending PING (ACK-of-ACK trigger) in packet 848
2021-03-19 22:14:55,146 DEBUG quic [b93fa38de016aed8] Received PING (probe) response
2021-03-19 22:14:55,154 DEBUG quic [b93fa38de016aed8] Sending PING (ACK-of-ACK trigger) in packet 856
2021-03-19 22:14:55,157 DEBUG quic [b93fa38de016aed8] Received PING (probe) response
2021-03-19 22:14:55,165 DEBUG quic [b93fa38de016aed8] Sending PING (ACK-of-ACK trigger) in packet 864
2021-03-19 22:14:55,167 DEBUG quic [b93fa38de016aed8] Received PING (probe) response
2021-03-19 22:14:55,175 DEBUG quic [b93fa38de016aed8] Sending PING (ACK-of-ACK trigger) in packet 872
2021-03-19 22:14:55,184 DEBUG quic [b93fa38de016aed8] Sending PING (ACK-of-ACK trigger) in packet 880
2021-03-19 22:14:55,188 DEBUG quic [b93fa38de016aed8] Received PING (probe) response
2021-03-19 22:14:55,194 DEBUG quic [b93fa38de016aed8] Sending PING (ACK-of-ACK trigger) in packet 888
2021-03-19 22:14:55,204 DEBUG quic [b93fa38de016aed8] Sending PING (ACK-of-ACK trigger) in packet 896
2021-03-19 22:14:55,206 DEBUG quic [b93fa38de016aed8] Received PING (probe) response
2021-03-19 22:14:55,213 DEBUG quic [b93fa38de016aed8] Sending PING (ACK-of-ACK trigger) in packet 904
2021-03-19 22:14:55,216 DEBUG quic [b93fa38de016aed8] Received PING (probe) response
2021-03-19 22:14:55,227 DEBUG quic [b93fa38de016aed8] Sending PING (ACK-of-ACK trigger) in packet 912
2021-03-19 22:14:55,237 DEBUG quic [b93fa38de016aed8] Local max_data raised to 33554432
2021-03-19 22:14:55,241 DEBUG quic [b93fa38de016aed8] Stream 0 local max_stream_data raised to 33554432
2021-03-19 22:14:55,242 DEBUG quic [b93fa38de016aed8] Sending PING (ACK-of-ACK trigger) in packet 920
2021-03-19 22:14:55,284 DEBUG quic [b93fa38de016aed8] Received PING (probe) response
2021-03-19 22:14:55,295 DEBUG quic [b93fa38de016aed8] Received PING (probe) response
2021-03-19 22:14:55,305 DEBUG quic [b93fa38de016aed8] Received PING (probe) response
2021-03-19 22:14:55,315 DEBUG quic [b93fa38de016aed8] Received PING (probe) response
2021-03-19 22:14:55,325 DEBUG quic [b93fa38de016aed8] Received PING (probe) response
2021-03-19 22:14:55,334 DEBUG quic [b93fa38de016aed8] Received PING (probe) response
2021-03-19 22:14:55,344 DEBUG quic [b93fa38de016aed8] Received PING (probe) response
2021-03-19 22:14:55,354 DEBUG quic [b93fa38de016aed8] Sending PING (ACK-of-ACK trigger) in packet 928
2021-03-19 22:14:55,355 DEBUG quic [b93fa38de016aed8] Received PING (probe) response
2021-03-19 22:14:55,363 DEBUG quic [b93fa38de016aed8] Sending PING (ACK-of-ACK trigger) in packet 936
2021-03-19 22:14:55,366 DEBUG quic [b93fa38de016aed8] Received PING (probe) response
2021-03-19 22:14:55,374 DEBUG quic [b93fa38de016aed8] Received PING (probe) response
2021-03-19 22:14:55,381 DEBUG quic [b93fa38de016aed8] Sending PING (ACK-of-ACK trigger) in packet 944
2021-03-19 22:14:55,384 DEBUG quic [b93fa38de016aed8] Received PING (probe) response
2021-03-19 22:14:55,393 DEBUG quic [b93fa38de016aed8] Received PING (probe) response
2021-03-19 22:14:55,403 DEBUG quic [b93fa38de016aed8] Received PING (probe) response
2021-03-19 22:14:55,413 DEBUG quic [b93fa38de016aed8] Received PING (probe) response
2021-03-19 22:14:55,427 DEBUG quic [b93fa38de016aed8] Received PING (probe) response
2021-03-19 22:14:55,437 DEBUG quic [b93fa38de016aed8] Sending PING (ACK-of-ACK trigger) in packet 952
2021-03-19 22:14:55,442 DEBUG quic [b93fa38de016aed8] Received PING (probe) response
2021-03-19 22:14:55,455 INFO client Response received for GET /8m_file : 8388608 bytes in 2.5 s (26.718 Mbps)
2021-03-19 22:14:55,456 DEBUG quic [b93fa38de016aed8] QuicConnectionState.CONNECTED -> QuicConnectionState.CLOSING
2021-03-19 22:14:56,135 DEBUG quic [b93fa38de016aed8] Discarding epoch Epoch.ONE_RTT
2021-03-19 22:14:56,136 DEBUG quic [b93fa38de016aed8] QuicConnectionState.CLOSING -> QuicConnectionState.TERMINATED
```
The difference of output I could see between the above two cases is that, during an "reduced-speed" file transfer, on the server side there's a KeyUnavailableError("Decryption key is not available"). Also, it appears that the client side doesn't send PING (probe) or PING (ACK-of-ACK trigger) in this case, which it always does in a "usual-speed" file transfer. But I haven't figured out if these behaviours are related to each other. Thanks for reading, I appreciate it. Any advises or explanation will be appreciated. Also pls let me know if you need more logs.

**Comment by jlaine** [comment]
This may be related to #129, there seem to be conditions where aioquic's retransmission is suboptimal. I'd like to get to the bottom of this!

In terms of trace a Wireshark capture (ideally on both client & server sides for the same test), but you'll also need to capture the encryption keys used with the `--secrets-log somefile.log` option.

**Comment by jlaine** [comment]
Another thing which probably doesn't help is that you seem to be having to negotiate the protocol version of each of your tests (so additional RTTs): are you running the same version of aioquic on client and server? I suspect the server is using an older version.

**Comment by maruibin** [comment]
Thank you so much for the reply! Sorry that I didn't update earlier. You're right - I was using an old version in my previous test. I switched to the lastest version and did four tests (normal vs keyUnavailableError; trasferring a 2MB and a 512kB file), please refer to the trace below.
[trace.zip](https://github.com/aiortc/aioquic/files/6201818/trace.zip)

I just realized after finishing the tests that the `--secrets-log` option appends lines to the file instead of overwriting it, so the keys I logged don't only include keys from that corresponding test, but also previous (possibly irrelavant) tests. So please refer to the keys in reverse order (since there're no timestamps to refer to), and please let me know if you do need a "clean" version of keys logs. So many thanks again!

**Comment by jlaine** [comment]
Hi @maruibin this should be fixed by #265 which triggers more agressive retransmission of CRYPTO data during the handshake. If you can still reproduce the issue with the latest code, could you please open a new issue with your new traces?

**Comment by Karthikdasari0423** [comment]
Hai ,

It is not an issue.
one small quick query

@maruibin
I am trying to upload file using aioquic(using below command) and not able to get how to do
and instead getting file with data as "Method Not Allowed" while using below command

I didnt made any changes to http3_client.py or http3_server.py

(python3 /root/aioquic/examples/http3_client.py --ca-certs /root/aioquic/tests/pycacert.pem -v https://1.2.3.4:4433/new.pdf --output-dir /tmp/ --data new.pdf)

any idea Do i need to place the file which i wanted to upload under any speific folder like /aioquic/examples/
(all my files are under /var/www/html/ directory)

Could you please help me to send file using aioquic http3_client.py "--data" argument.

Could you please help me to transfer files between an http/3 client and an http/3 server.

Could you please confirm were you able to create multiple streams in a single connection using the AIOQUIC code .

Looking forward for your reply

Thank you.
